src/Mod/PartDesign/App/FeatureHole.cpp

```diff
@@ -2084,34 +2345,70 @@ TopoDS_Shape Hole::makeThread(const gp_Vec& xDir, const gp_Vec& zDir, double len
+        clearance = getThreadClassClearance() / 2;
 
     BRepBuilderAPI_MakeWire mkThreadWire;
-    mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p1, p2).Edge());
-    mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p2, p3).Edge());
-    mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p3, p4).Edge());
-    mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p4, p5).Edge());
-    mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p5, p6).Edge());
-    mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p6, p1).Edge());
+    double H;
+    std::string threadTypeStr = ThreadType.getValueAsString();
+    if (threadTypeStr == "BSP" || threadTypeStr == "BSW" || threadTypeStr == "BSF") {
+        H = 0.960491 * Pitch; // Height of Sharp V
+        double radius  = 0.137329 * Pitch; // radius of the crest
+        // construct the cross section going counter-clockwise
+        // --------------
+        // P    | p4
+        // 5/8P |                p3
+        //      |                         crest
+        // 3/8P |                p2
+        // 0    | p1
+        // --------------
+        //      | base-sharpV    9/16r    dmaj     H
+
+        gp_Pnt p1 = toPnt((Dmaj - 5 * H / 6 + clearance) * xDir + .001 * Pitch * zDir);
+        gp_Pnt p4 = toPnt((Dmaj - 5 * H / 6 + clearance) * xDir + .999 * Pitch * zDir);
+
+        // Calculate positions for p2 and p3 based on the arc radius
+        double p23x = Dmaj - 9 * radius / 16 + clearance;
+
+        gp_Pnt p2 = toPnt(p23x * xDir + 3 * Pitch / 8 * zDir);
+        gp_Pnt p3 = toPnt(p23x * xDir + 5 * Pitch / 8 * zDir);
+        gp_Pnt crest = toPnt((Dmaj + clearance) * xDir + Pitch / 2 * zDir);
+
+        mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p1, p2).Edge());
+        Handle(Geom_TrimmedCurve) arc1 = GC_MakeArcOfCircle(p2, crest, p3).Value();
+        mkThreadWire.Add(BRepBuilderAPI_MakeEdge(arc1).Edge());
+        mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p3, p4).Edge());
+        mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p4, p1).Edge());
+    } else {
+        H = sqrt(3) / 2 * Pitch; // height of fundamental triangle
+        double h = 7 * H / 8; // distance from Dmaj to the base
+        // construct the cross section going counter-clockwise
+        // --------------
+        // P     | p4
+        // 9/16P |                p3
+        // 7/16P |                p2
+        // 0     | p1
+        // --------------
+        //       | base-sharpV    dmaj
+
+        gp_Pnt p1 = toPnt((Dmaj - h + clearance) * xDir + .001 * zDir);
+        gp_Pnt p2 = toPnt((Dmaj + clearance) * xDir + 7 * Pitch / 16 * zDir);
+        gp_Pnt p3 = toPnt((Dmaj + clearance) * xDir + 9 * Pitch / 16 * zDir);
+        gp_Pnt p4 = toPnt((Dmaj - h + clearance) * xDir + .999 * Pitch * zDir);
+
+        mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p1, p2).Edge());
+        mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p2, p3).Edge());
+        mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p3, p4).Edge());
+        mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p4, p1).Edge());

```

Comment:
Is there really a need to change the ISO/UTS profile when introducing Whitworth? What is the motivation for this change? Is it to prevent residuals when producing tapered threads? Is it to increase performance (fewer points)? 

The original profile had six points -- [it extended the trapezium with a small rectangle](https://forum.freecad.org/viewtopic.php?f=19&t=54284#p466570) (presumably to prevent residuals in the boolean cut due to inaccuracies, but that's guess work). Your profile does not do this. Why did you remove it?

The original profile had exact flank angles of 30°. Yours is slightly squeezed due to the 0.001/0.999 zDir, giving a less accurate profile with 29.94°. Why? Is it to prevent self-intersections? Then I'd propose to reduce it in xDir as well (by twice the amount of your z-squeeze, aka by 1/sin(30°)) to keep the flank angle.

Reply:
> Is there really a need to change the ISO/UTS profile when introducing Whitworth? What is the motivation for this change? Is it to prevent residuals when producing tapered threads? Is it to increase performance (fewer points)?
> The original profile had six points -- [it extended the trapezium with a small rectangle](https://forum.freecad.org/viewtopic.php?f=19&t=54284#p466570) (presumably to prevent residuals in the boolean cut due to inaccuracies, but that's guess work). Your profile does not do this. Why did you remove it?

yes, exactly both things:
- p1 and p4 in that link/picture are a problem with tapered, they need to go deeper.
- having 6 points is slower

> The original profile had exact flank angles of 30°. Yours is slightly squeezed due to the 0.001/0.999 zDir, giving a less accurate profile with 29.94°. Why? Is it to prevent self-intersections? Then I'd propose to reduce it in xDir as well (by twice the amount of your z-squeeze, aka by 1/sin(30°)) to keep the flank angle.

yes, workaround due to those issues, if i understood correctly it's "OCCT Confusion"
i'll take a look into it, thanks for the suggestion